File: Tree/BinaryTree/C++/binaryTree.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <string>

using namespace std;

/// ======================= TreeNode =======================

/**
 * @class TreeNode
 * @brief Represents a node in a binary tree.
 * @tparam T Type of the data stored in the node
 */
template<typename T>
class TreeNode {
public:
    T data;            ///< Value stored in the node
    TreeNode* left;    ///< Pointer to the left child node
    TreeNode* right;   ///< Pointer to the right child node

    /**
     * @brief Constructor: Creates a new node with the given value.
     * @param value Value to store in the node
     */
    TreeNode(T value) : data(value), left(nullptr), right(nullptr) {}
};

/// ======================= BinaryTree =======================

/**
 * @class BinaryTree
 * @brief Template class representing a binary tree with level-order insertion,
 *        traversals, deletion, and ASCII visual printing.
 * @tparam T Type of data stored in the tree nodes
 */
template<typename T>
class BinaryTree {
private:
    TreeNode<T>* root; ///< Pointer to the root node of the tree

public:
    /**
     * @brief Constructs an empty binary tree.
     */
    BinaryTree() : root(nullptr) {}

    // -------------------- Insert (Level-Order) --------------------
    /**
     * @brief Inserts a new node with the specified value into the tree using level-order.
     * @param value Value to insert into the tree
     */
    void insert(T value) {
        TreeNode<T>* newNode = new TreeNode<T>(value);
        if (!root) { root = newNode; return; }

        queue<TreeNode<T>*> q;
        q.push(root);

        while (!q.empty()) {
            TreeNode<T>* current = q.front(); q.pop();
            if (!current->left) { current->left = newNode; return; } else q.push(current->left);
            if (!current->right) { current->right = newNode; return; } else q.push(current->right);
        }
    }

    // -------------------- Height --------------------
    /**
     * @brief Computes the height of a subtree starting from the given node.
     * @param node Node to start computing height from
     * @return Height of the subtree
     */
    int height(TreeNode<T>* node) {
        if (!node) return 0;
        return 1 + max(height(node->left), height(node->right));
    }

    /**
     * @brief Computes the height of the entire tree.
     * @return Height of the tree
     */
    int height() { return height(root); }

    // -------------------- Pre-Order Traversal --------------------
    void preOrder(TreeNode<T>* node, vector<T>& result) {
        if (!node) return;
        result.push_back(node->data);
        preOrder(node->left, result);
        preOrder(node->right, result);
    }

    vector<T> preOrder() {
        vector<T> result;
        preOrder(root, result);
        return result;
    }

    // -------------------- In-Order Traversal --------------------
    void inOrder(TreeNode<T>* node, vector<T>& result) {
        if (!node) return;
        inOrder(node->left, result);
        result.push_back(node->data);
        inOrder(node->right, result);
    }

    vector<T> inOrder() {
        vector<T> result;
        inOrder(root, result);
        return result;
    }

    // -------------------- Post-Order Traversal --------------------
    void postOrder(TreeNode<T>* node, vector<T>& result) {
        if (!node) return;
        postOrder(node->left, result);
        postOrder(node->right, result);
        result.push_back(node->data);
    }

    vector<T> postOrder() {
        vector<T> result;
        postOrder(root, result);
        return result;
    }

    // -------------------- Find Node --------------------
    TreeNode<T>* find(T value, TreeNode<T>* node) {
        if (!node) return nullptr;
        if (node->data == value) return node;

        TreeNode<T>* leftRes = find(value, node->left);
        if (leftRes) return leftRes;

        return find(value, node->right);
    }

    TreeNode<T>* find(T value) { return find(value, root); }

    // -------------------- Delete Node --------------------
    void deleteNode(T value) {
        TreeNode<T>* target = find(value);
        if (!target) throw runtime_error("Node does not exist.");

        TreeNode<T>* lastRight = root;
        TreeNode<T>* parent = nullptr;

        while (lastRight->right) { parent = lastRight; lastRight = lastRight->right; }

        target->data = lastRight->data;

        if (parent == nullptr) root = nullptr;
        else parent->right = nullptr;

        delete lastRight;
    }

    // -------------------- Print --------------------
    struct NodeInfo {
        string text;           ///< Node's text
        int width;             ///< Width of the subtree
        int middle;            ///< Middle position for connecting branches
        vector<string> lines;  ///< Lines representing the subtree
    };

    NodeInfo buildTree(TreeNode<T>* node) {
        if (!node) return {"", 0, 0, {}};

        string val = string(1, node->data); // direct conversion for char or string assumed to be T

        NodeInfo left = buildTree(node->left);
        NodeInfo right = buildTree(node->right);

        int width = left.width + val.length() + right.width;
        int middle = left.width + val.length() / 2;

        vector<string> lines;

        // First line: node value
        lines.push_back(string(left.width, ' ') + val + string(right.width, ' '));

        // Second line: branches
        string branch(width, ' ');
        if (!left.lines.empty()) branch[left.width - 1] = '/';
        if (!right.lines.empty()) branch[left.width + val.length()] = '\\';
        lines.push_back(branch);

        // Merge children
        int height = max(left.lines.size(), right.lines.size());
        for (int i = 0; i < height; i++) {
            string L = i < left.lines.size() ? left.lines[i] : string(left.width, ' ');
            string R = i < right.lines.size() ? right.lines[i] : string(right.width, ' ');
            lines.push_back(L + string(val.length(), ' ') + R);
        }

        return {val, width, middle, lines};
    }

    void print() {
        if (!root) return;
        NodeInfo tree = buildTree(root);
        for (auto& line : tree.lines) cout << line << endl;
    }
};

File: Tree/BinaryTree/Node.js/binaryTree.js
/**
 * Class representing a node in a binary tree.
 */
class TreeNode {
    /**
     * Constructs a new TreeNode with the given value.
     * @param {any} data - The value to store in the node.
     */
    constructor(data) {
        /** @type {any} The value stored in the node */
        this.data = data;

        /** @type {TreeNode|null} Pointer to the left child node */
        this.left = null;

        /** @type {TreeNode|null} Pointer to the right child node */
        this.right = null;
    }
}

/**
 * Class representing a Binary Tree (not necessarily a BST).
 */
module.exports = class BinaryTree {
    /** @type {TreeNode|null} Root node of the binary tree */
    #root;

    /**
     * Constructs an empty BinaryTree.
     */
    constructor() {
        this.#root = null;
    }

    /**
     * Inserts a new value into the tree following level-order (BFS) insertion.
     * @param {any} data - The value to insert into the tree.
     */
    insert(data) {
        const newNode = new TreeNode(data);

        if (!this.#root) {
            this.#root = newNode;
            return;
        }

        const queue = [this.#root];
        while (queue.length) {
            const current = queue.shift();

            if (!current.left) {
                current.left = newNode;
                return;
            } else {
                queue.push(current.left);
            }

            if (!current.right) {
                current.right = newNode;
                return;
            } else {
                queue.push(current.right);
            }
        }
    }

    /**
     * Calculates the height of the tree or a subtree.
     * @param {TreeNode} [node=this.#root] - The node to calculate height from.
     * @returns {number} The height of the tree.
     */
    height(node = this.#root) {
        if (!node) return 0;
        return 1 + Math.max(this.height(node.left), this.height(node.right));
    }

    /**
     * Performs a pre-order traversal of the tree.
     * @param {TreeNode} [node=this.#root] - Node to start traversal from.
     * @param {any[]} [result=[]] - Array to store traversal result.
     * @returns {any[]} Array of node values in pre-order.
     */
    preOrder(node = this.#root, result = []) {
        if (!node) return result;

        result.push(node.data);
        this.preOrder(node.left, result);
        this.preOrder(node.right, result);

        return result;
    }

    /**
     * Performs an in-order traversal of the tree.
     * @param {TreeNode} [node=this.#root] - Node to start traversal from.
     * @param {any[]} [result=[]] - Array to store traversal result.
     * @returns {any[]} Array of node values in in-order.
     */
    inOrder(node = this.#root, result = []) {
        if (!node) return result;

        this.inOrder(node.left, result);
        result.push(node.data);
        this.inOrder(node.right, result);

        return result;
    }

    /**
     * Performs a post-order traversal of the tree.
     * @param {TreeNode} [node=this.#root] - Node to start traversal from.
     * @param {any[]} [result=[]] - Array to store traversal result.
     * @returns {any[]} Array of node values in post-order.
     */
    postOrder(node = this.#root, result = []) {
        if (!node) return result;

        this.postOrder(node.left, result);
        this.postOrder(node.right, result);
        result.push(node.data);

        return result;
    }

    /**
     * Finds a node with the given value in the tree.
     * @param {any} data - Value to search for.
     * @param {TreeNode} [node=this.#root] - Node to start search from.
     * @returns {TreeNode|null} The node with the given value or null if not found.
     */
    find(data, node = this.#root) {
        if (!node) return null;
        if (node.data === data) return node;

        const leftResult = this.find(data, node.left);
        if (leftResult) return leftResult;

        return this.find(data, node.right);
    }

    /**
     * Deletes a node with the specified value from the tree.
     * Replaces the target node's value with the rightmost node's value, then removes the rightmost node.
     * @param {any} data - Value of the node to delete.
     * @throws {Error} If the node does not exist.
     */
    delete(data) {
        let targetNode = this.find(data);

        if (!targetNode) {
            throw new Error("Node does not exists.");
        } else {
            let lastRightNode = this.#root;
            let parent = null;
            while (lastRightNode.right) {
                parent = lastRightNode;
                lastRightNode = lastRightNode.right;
            }
            targetNode.data = lastRightNode.data;

            if (parent === null) {
                this.#root = null;
            } else {
                parent.right = null;
            }
        }
    }

    /**
     * Prints the binary tree in a visual tree-like structure.
     */
    print() {
        if (!this.#root) return;

        const buildTree = (node) => {
            if (!node) return { lines: [], width: 0, height: 0, middle: 0 };

            const valStr = node.data.toString();
            const left = buildTree(node.left);
            const right = buildTree(node.right);

            const height = Math.max(left.height, right.height) + 2;
            const width = left.width + valStr.length + right.width;
            const middle = left.width + Math.floor(valStr.length / 2);

            const lines = [];

            // Node value line
            let firstLine = " ".repeat(left.width) + valStr + " ".repeat(right.width);
            lines.push(firstLine);

            // Branches line
            let secondLine = "";
            if (left.height > 0) {
                secondLine += " ".repeat(left.width - 1) + "/" + " ".repeat(valStr.length + right.width);
            } else {
                secondLine += " ".repeat(left.width + valStr.length + right.width);
            }

            if (right.height > 0 && left.height === 0) {
                secondLine = secondLine.slice(0, left.width + valStr.length) + "\\" + secondLine.slice(left.width + valStr.length + 1);
            } else if (right.height > 0) {
                secondLine = secondLine.slice(0, left.width + valStr.length + right.width - right.width) + "\\" + secondLine.slice(left.width + valStr.length + right.width - right.width + 1);
            }

            lines.push(secondLine);

            // Merge left and right subtree lines
            for (let i = 0; i < Math.max(left.height, right.height); i++) {
                const leftLine = i < left.lines.length ? left.lines[i] : " ".repeat(left.width);
                const rightLine = i < right.lines.length ? right.lines[i] : " ".repeat(right.width);
                lines.push(leftLine + " ".repeat(valStr.length) + rightLine);
            }

            return { lines, width, height, middle };
        };

        const { lines } = buildTree(this.#root);
        for (const line of lines) {
            console.log(line);
        }
    }
};
